pkg/ebpf/generic_map_test.go

```diff
@@ -0,0 +1,518 @@
+	for it.Next(&k, &entries) {
+		numElements++
+		foundElements[k] = true
+
+		for i := 0; i < nbCpus; i++ {
+			require.Equal(t, k+uint32(i), entries[i])
+		}
+	}
+
+	require.Equal(t, len(numsToPut), numElements)
+	for _, num := range numsToPut {
+		require.True(t, foundElements[num])
+	}
+}
+
+type ValueStruct struct {
+	A uint32
+	B uint32
+}
+
+func TestIterateWithValueStructs(t *testing.T) {
+	doTest := func(t *testing.T, batchSize int, oneItem bool) {
+		singleItem := batchSize == 1
+		if !singleItem && !BatchAPISupported() {
+			t.Skip("Batch API not supported")
+		}
+
+		m, err := NewGenericMap[uint32, ValueStruct](&ebpf.MapSpec{
+			Type:       ebpf.Hash,
+			MaxEntries: 10,
+		})
+		require.NoError(t, err)
+
+		var numsToPut []uint32
+		if oneItem {
+			numsToPut = []uint32{10}
+		} else {
+			numsToPut = []uint32{0, 100, 200, 300, 400, 500}
+		}
+		for _, num := range numsToPut {
+			v := ValueStruct{A: num, B: num + 1}
+			require.NoError(t, m.Put(&num, &v))
+		}
+
+		var k uint32
+		var v ValueStruct
+		numElements := 0
+		foundElements := make(map[uint32]bool)
+
+		it := m.IterateWithOptions(batchSize)
+		require.NotNil(t, it)
+		if singleItem {
+			require.IsType(t, &genericMapItemIterator[uint32, ValueStruct]{}, it)
+		} else {
+			require.IsType(t, &genericMapBatchIterator[uint32, ValueStruct]{}, it)
+		}
+
+		for it.Next(&k, &v) {
+			numElements++
+			foundElements[k] = true
+
+			require.Equal(t, k, v.A)
+			require.Equal(t, k+1, v.B)
+		}
+
+		require.Equal(t, len(numsToPut), numElements)
+		for _, num := range numsToPut {
+			require.True(t, foundElements[num])
+		}
+	}
+
+	t.Run("SingleItem", func(t *testing.T) {
+		doTest(t, 1, false)
+	})
+
+	t.Run("Batch", func(t *testing.T) {
+		doTest(t, 0, false)
+	})
+
+	t.Run("BatchWithOneItem", func(t *testing.T) {
+		doTest(t, 0, true)
+	})
+}
+
+func TestBatchIterAllocsPerRun(t *testing.T) {
+	if !BatchAPISupported() {
+		t.Skip("Batch API not supported")
+	}
+
+	m, err := NewGenericMap[uint32, uint32](&ebpf.MapSpec{
+		Type:       ebpf.Hash,
+		MaxEntries: 10000,
+	})
+	require.NoError(t, err)
+
+	numsToPut := uint32(9000)
+	for i := uint32(0); i < numsToPut; i++ {
+		require.NoError(t, m.Put(&i, &i))
+	}
+
+	var k uint32
+	var v uint32
+	batchSize := 10
+
+	allocsSmallBatch := testing.AllocsPerRun(100, func() {
+		numElements := uint32(0)
+		it := m.IterateWithOptions(batchSize)
+		for it.Next(&k, &v) {
+			numElements++
+		}
+		require.Equal(t, numsToPut, numElements)
+	})
+
+	batchSize = 100
+
+	allocsLargerBatch := testing.AllocsPerRun(100, func() {
+		numElements := uint32(0)
+		it := m.IterateWithOptions(batchSize)
+		for it.Next(&k, &v) {
+			numElements++
+		}
+		require.Equal(t, numsToPut, numElements)
+	})
+
+	require.LessOrEqual(t, allocsSmallBatch, 8.0)
+	require.LessOrEqual(t, allocsLargerBatch, 8.0)
+	require.Equal(t, allocsLargerBatch, allocsSmallBatch) // We don't want allocations to depend on batch size
```

Comment:
We just want to check that iterating doesn't do any additional allocations. It is expected that different batch sizes will have different allocation costs up front.
```suggestion
	it := m.IterateWithOptions(batchSize)
	allocs := testing.AllocsPerRun(1, func() {
		it.Next(&k, &v)
	})
	assert.EqualValues(t, allocs, 0)

	batchSize = 100
	it = m.IterateWithOptions(batchSize)
	allocs = testing.AllocsPerRun(1, func() {
		it.Next(&k, &v)
	})
	assert.EqualValues(t, allocs, 0)
```